lib/src/fix.rs

```diff
@@ -0,0 +1,319 @@
+            || updates_tx.clone(),
+            |updates_tx, file_to_fix| -> Result<(), FixError> {
+                let result = (self.fix_fn)(store, file_to_fix)?;
+                match result {
+                    Some(new_file_id) => {
+                        updates_tx.send((file_to_fix, new_file_id)).unwrap();
+                        Ok(())
+                    }
+                    None => Ok(()),
+                }
+            },
+        )?;
+        drop(updates_tx);
+        let mut result = HashMap::new();
+        while let Ok((file_to_fix, new_file_id)) = updates_rx.recv() {
+            result.insert(file_to_fix, new_file_id);
+        }
+        Ok(result)
+    }
+}
+
+/// Updates files with formatting fixes or other changes, using the given
+/// FileFixer.
+///
+/// The primary use case is to apply the results of automatic code formatting
+/// tools to revisions that may not be properly formatted yet. It can also be
+/// used to modify files with other tools like `sed` or `sort`.
+///
+/// After the FileFixer is done, descendants are also updated, which ensures
+/// that the fixes are not lost. This will never result in new conflicts. Files
+/// with existing conflicts are updated on all sides of the conflict, which
+/// can potentially increase or decrease the number of conflict markers.
+pub fn fix_files(
+    root_commits: Vec<CommitId>,
+    matcher: Box<dyn Matcher>,
+    include_unchanged_files: bool,
+    repo_mut: &mut MutableRepo,
+    file_fixer: &impl FileFixer,
+) -> Result<FixSummary, FixError> {
+    let mut summary = FixSummary {
+        rewrites: HashMap::new(),
+        num_checked_commits: 0,
+        num_fixed_commits: 0,
+    };
+
+    // Collect all of the unique `FileToFix`s we're going to use. file_fixer should
+    // be deterministic, and should not consider outside information, so it is
+    // safe to deduplicate inputs that correspond to multiple files or commits.
+    // This is typically more efficient, but it does prevent certain use cases
+    // like providing commit IDs as inputs to be inserted into files. We also
+    // need to record the mapping between files-to-fix and paths/commits, to
+    // efficiently rewrite the commits later.
+    //
+    // If a path is being fixed in a particular commit, it must also be fixed in all
+    // that commit's descendants. We do this as a way of propagating changes,
+    // under the assumption that it is more useful than performing a rebase and
+    // risking merge conflicts. In the case of code formatters, rebasing wouldn't
+    // reliably produce well formatted code anyway. Deduplicating inputs helps
+    // to prevent quadratic growth in the number of tool executions required for
+    // doing this in long chains of commits with disjoint sets of modified files.
+    let commits: Vec<_> = RevsetExpression::commits(root_commits.clone())
+        .descendants()
+        .evaluate(repo_mut.base_repo().as_ref())?
```

Comment:
nit: perhaps we have to use `repo_mut` here because we can no longer assume that the current `repo_mut` has no mutation.

Reply:
I am not sure I understand your comment. Are you saying I should change the logic? Or add some comment?